internal/cache/file/downloader/job.go

```diff
@@ -226,47 +225,138 @@ func (job *Job) notifySubscribers() {
+// Acquires and releases Lock(job.mu)
+func (job *Job) updateStatusAndNotifySubscribers(statusName jobStatusName, statusErr error) {
+	if statusName == Failed {
+		logger.Errorf("Job:%p (%s:/%s) Failed with error: %v", job, job.bucket.Name(), job.object.Name, statusErr)
+	} else {
+		logger.Tracef("Job:%p (%s:/%s) status changed to %v with error: %v", job, job.bucket.Name(), job.object.Name, statusName, statusErr)
+	}
 	job.mu.Lock()
-	job.status.Err = downloadErr
-	job.status.Name = Failed
+	job.status.Err = statusErr
+	job.status.Name = statusName
 	job.notifySubscribers()
 	job.mu.Unlock()
 }
 
-// updateFileInfoCache updates the file info cache with latest offset downloaded
-// by job. Returns error in case of failure.
-//
-// Not concurrency safe and requires LOCK(job.mu)
-func (job *Job) updateFileInfoCache() (err error) {
+// updateStatusOffset updates the offset in job's status and in file info cache
+// with the given offset. If the the update is successful, this function also
+// notify the subscribers.
+func (job *Job) updateStatusOffset(downloadedOffset int64) (err error) {
 	fileInfoKey := data.FileInfoKey{
 		BucketName: job.bucket.Name(),
 		ObjectName: job.object.Name,
 	}
 	fileInfoKeyName, err := fileInfoKey.Key()
 	if err != nil {
-		err = fmt.Errorf("updateFileInfoCache: error while creating fileInfoKeyName for bucket %s and object %s %w",
+		err = fmt.Errorf("updateStatusOffset: error while creating fileInfoKeyName for bucket %s and object %s %w",
 			fileInfoKey.BucketName, fileInfoKey.ObjectName, err)
-		return
+		return err
 	}
 
 	updatedFileInfo := data.FileInfo{
 		Key: fileInfoKey, ObjectGeneration: job.object.Generation,
-		FileSize: job.object.Size, Offset: uint64(job.status.Offset),
+		FileSize: job.object.Size, Offset: uint64(downloadedOffset),
 	}
-
-	logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+	job.mu.Lock()
+	defer job.mu.Unlock()
 	err = job.fileInfoCache.UpdateWithoutChangingOrder(fileInfoKeyName, updatedFileInfo)
-	if err != nil {
-		err = fmt.Errorf("updateFileInfoCache: error while inserting into fileInfoCache %s: %w", updatedFileInfo.Key, err)
-		return
+	if err == nil {
+		job.status.Offset = downloadedOffset
+		// Notify subscribers if file cache is updated.
+		logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+		job.notifySubscribers()
+		return err
 	}
-	return
+
+	err = fmt.Errorf("updateStatusOffset: error while updating offset: %v in fileInfoCache %s: %w", downloadedOffset, updatedFileInfo.Key, err)
+	return err
+}
+
+// downloadObjectToFile downloads the backing object from GCS into the given
+// file and updates the file info cache. It uses gcs.Bucket's NewReader method
+// to download the object.
+func (job *Job) downloadObjectToFile(cacheFile *os.File) (err error) {
```

Comment:
Are we writing unit test for this?

Reply:
Yes, will be there in next PR. Not a blocker for this PR since we already have unit tests for Download function and didn't want to make the PR longer.